### PR TITLE
[MRG+1] Deprecate is_string_like, is_sequence_of_strings

### LINKED ISSUE #7725: `is_string_like` returns True for numpy `object` arrays
### Bug report

**Bug summary**

Function `is_string_like` returns True for numpy `object` arrays

**Code for reproduction**

```python
import numpy as np
from matplotlib.cbook import is_string_like
print(is_string_like(np.array(map(str, [1,2,3]), dtype=object)))
```

**Actual outcome**
```
True
```

**Expected outcome**
```
False
```

The problem is that the line 707 `obj + ''` in `is_string_like` does not generate an exception with such numpy array.

**Matplotlib version**

- [ ] Matplotlib version: 1.5.3
- [ ] OS: Ubuntu 14.04
- [ ] Python and Matplotlib Installed with Anaconda

### PR DESCRIPTION
**Pull request by anntzer**
While implementing #7996 (see also #7835), I had missed that numpy's `str_` (and `unicode_` on Py2) derive from the corresponding builtin type(s), so `isinstance(x, six.text_type)` is just as good.  So I just inlined that and restored the old definition of `is_string_like`, immediately deprecating it.

Would also (upon removal) close #7725.

### CODE REVIEW

**Comment by QuLogic** [comment]
Has this been true since NumPy 1.7.1?

**Comment by anntzer** [comment]
Yes.
```
$ conda create -n tmpenv --yes python=3.3 numpy=1.7.1 >/dev/null && source activate tmpenv &&
python -c 'from numpy import str_, unicode_; print(str in str_.__mro__, str in unicode_.__mro__)'
True True
```
and likewise on Py2 (but with `unicode in unicode.__mro__`)

**Comment by QuLogic** [comment]
Looks like an example might have broken.

**Comment by anntzer** [comment]
Fixed.
`Axes3D.scatter` now doesn't accept 2D inputs anymore (which were silently flattened before).  If we want to document the change it should probably go with #7786.

**Comment by anntzer** [comment]
rebased.  bumping :-)

File: doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
+`cbook.is_numlike` only performs an instance check, `cbook.is_string_like` is deprecated
+````````````````````````````````````````````````````````````````````````````````````````
 
-`cbook.is_numlike` and `cbook.is_string_like` now only check that
-their argument is an instance of ``(numbers.Number, np.Number)`` and
-``(six.string_types, np.str_, np.unicode_)`` respectively.  In particular, this
-means that arrays are now never num-like or string-like regardless of their
-dtype.
+`cbook.is_numlike` now only checks that its argument is an instance of
+``(numbers.Number, np.Number)`` and.  In particular, this means that arrays are
```
Comment (QuLogic) [inline doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst]: and what?

File: examples/mplot3d/2dcollections3d.py [hunk 2]
```diff
@@ -25,11 +25,11 @@
...
 np.random.seed(19680801)
 
-x = np.random.sample(20*len(colors))
-y = np.random.sample(20*len(colors))
+x = np.random.sample(20 * len(colors))
+y = np.random.sample(20 * len(colors))
 c_list = []
 for c in colors:
-    c_list.append([c]*20)
+    c_list.extend([c] * 20)
```
Comment (QuLogic) [inline examples/mplot3d/2dcollections3d.py]: Is this actually related?

File: doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst [hunk 3]
```diff
@@ -1,8 +1,10 @@
...
+now not num-like.
+
+`cbook.is_string_like` and `cbook.is_sequence_of_strings` have been
+deprecated.  Use ``isinstance(obj, six.string_types)`` and ``all(isinstance(o,
+six.string_types) for o in obj) and not iterable(obj)`` instead.
```
Comment (QuLogic) [inline doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst]: That doesn't seem right:
```python
In [9]: from matplotlib.cbook import iterable, is_sequence_of_strings

In [10]: obj = ['a', 'b', 'c']

In [11]: is_sequence_of_strings(obj)
Out[11]: True

In [12]: import six

In [13]: all(isinstance(o, six.string_types) for o in obj) and not iterable(obj)
Out[13]: False
```
(using 2.0.0, though)

File: lib/matplotlib/cbook/__init__.py [hunk 4]
```diff
@@ -1590,13 +1592,14 @@ def delete_masked_points(*args):
...
         return ()
-    if (is_string_like(args[0]) or not iterable(args[0])):
+    if (isinstance(args[0], six.string_types) or not iterable(args[0])):
         raise ValueError("First argument must be a sequence")
     nrecs = len(args[0])
     margs = []
     seqlist = [False] * len(args)
     for i, x in enumerate(args):
-        if (not is_string_like(x)) and iterable(x) and len(x) == nrecs:
+        if ((not isinstance(x, six.string_types)) and iterable(x)
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py]: Can drop extra parentheses.

File: lib/matplotlib/cbook/__init__.py [hunk 5]
```diff
@@ -1219,10 +1221,10 @@ def finddir(o, match, case=False):
     is True require an exact case match.
     """
     if case:
-        names = [(name, name) for name in dir(o) if is_string_like(name)]
+        names = [(name, name) for name in dir(o) if isinstance(name, six.string_types)]
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py]: Isn't this too long?

File: lib/matplotlib/colors.py [hunk 6]
```diff
@@ -789,7 +789,7 @@ def __init__(self, colors, name='from_list', N=None):
         if N is None:
             N = len(self.colors)
         else:
-            if (cbook.is_string_like(self.colors) and
+            if (isinstance(self.colors, six.string_types) and
                     cbook.is_hashable(self.colors)):
```
Comment (QuLogic) [inline lib/matplotlib/colors.py]: Hmm, what strings are not hashable?

File: lib/matplotlib/lines.py [hunk 7]
```diff
@@ -1289,7 +1289,7 @@ def update_from(self, other):
 
     def _get_rgba_face(self, alt=False):
         facecolor = self._get_markerfacecolor(alt=alt)
-        if is_string_like(facecolor) and facecolor.lower() == 'none':
+        if isinstance(facecolor, six.string_types) and facecolor.lower() == 'none':
```
Comment (QuLogic) [inline lib/matplotlib/lines.py]: Probably too long, also.

File: lib/matplotlib/text.py [hunk 8]
```diff
@@ -1865,14 +1865,14 @@ def _get_ref_xy(self, renderer):
 
         if isinstance(self.xycoords, tuple):
             s1, s2 = self.xycoords
-            if ((is_string_like(s1) and s1.split()[0] == "offset") or
-                  (is_string_like(s2) and s2.split()[0] == "offset")):
+            if ((isinstance(s1, six.string_types) and s1.split()[0] == "offset") or
+                  (isinstance(s2, six.string_types) and s2.split()[0] == "offset")):
```
Comment (QuLogic) [inline lib/matplotlib/text.py]: Looks a bit over as well.

File: doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst (hunk 1 above)
Comment (anntzer) [inline doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst]: fixed

File: doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst (hunk 3 above)
Comment (anntzer) [inline doc/api/api_changes/2017-01-30-AL_is_numlike_stringlike.rst]: the "not" needs to be removed, fixed

File: lib/matplotlib/text.py (hunk 8 above)
Comment (anntzer) [inline lib/matplotlib/text.py]: fixed

File: lib/matplotlib/colors.py (hunk 6 above)
Comment (anntzer) [inline lib/matplotlib/colors.py]: probably was the case of masked string arrays.  removed the now redundant check.

File: lib/matplotlib/cbook/__init__.py (hunk 4 above)
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: fixed

File: lib/matplotlib/cbook/__init__.py (hunk 5 above)
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: fixed

File: examples/mplot3d/2dcollections3d.py (hunk 2 above)
Comment (anntzer) [inline examples/mplot3d/2dcollections3d.py]: not really, moved out.

File: lib/matplotlib/lines.py (hunk 7 above)
Comment (anntzer) [inline lib/matplotlib/lines.py]: fixed

File: examples/mplot3d/2dcollections3d.py (hunk 2 above)
Comment (anntzer) [inline examples/mplot3d/2dcollections3d.py]: oh actually i remember now, it is.
this is because 3d scatter no longer implicitly flattens its input (see the last hunk of the diff).

Comment (QuLogic) [inline examples/mplot3d/2dcollections3d.py]: Probably should add that to the API what's new, then.

File: lib/matplotlib/colors.py (hunk 6 above)
Comment (QuLogic) [inline lib/matplotlib/colors.py]: I think there might be a few more of these that I didn't comment on.

File: examples/mplot3d/2dcollections3d.py (hunk 2 above)
Comment (anntzer) [inline examples/mplot3d/2dcollections3d.py]: Done.  Also documented the similar change in #7786 at the same time.

File: lib/matplotlib/colors.py (hunk 6 above)
Comment (anntzer) [inline lib/matplotlib/colors.py]: Caught two more, should be all fixed now.

Comment (QuLogic) [inline lib/matplotlib/colors.py]: I don't see it...

Comment (anntzer) [inline lib/matplotlib/colors.py]: https://github.com/matplotlib/matplotlib/pull/8011/commits/c773d80057944bcb4f387a7f1ec87e71deeac2d1

Comment (QuLogic) [inline lib/matplotlib/colors.py]: There's only one line that changes `is_hashable` there (this one.)

Comment (anntzer) [inline lib/matplotlib/colors.py]: Ah sorry, thought you meant overly long lines.  Now I have removed the hashability checks too.

File: lib/matplotlib/blocking_input.py [hunk 9]
```diff
@@ -40,8 +39,6 @@ class BlockingInput(object):
     """
     def __init__(self, fig, eventslist=()):
         self.fig = fig
-        if not is_sequence_of_strings(eventslist):
-            raise ValueError("Requires a sequence of event name strings")
```
Comment (tacaswell) [inline lib/matplotlib/blocking_input.py]: Why remove this?

Comment (anntzer) [inline lib/matplotlib/blocking_input.py]: Because it's not as if the check was really useful anyways -- one can always pass a sequence of invalid event names and it'll be just as bad as of one passed in a sequence of nonstrings.
